### PR TITLE
Issue with canceling XHR request using promises - fixed #12525

### LINKED ISSUE #10890: $resource does not abort the requests through the promise
https://github.com/angular/angular.js/blob/master/src/ngResource/resource.js#L564

### PR DESCRIPTION
**Pull request by netman92**
Fixes #10890 and #9332.

@jtheoof no conflicts, is it ok now?

### CODE REVIEW

**Comment by jtheoof** [comment]
:+1: Although the commit message does not seem to follow the angular commit message conventions. I would preprend the commit the message with: `fix(ngResource): ...`

**Comment by netman92** [comment]
@jtheoof fixed

File: src/ngResource/resource.js [hunk 1]
```diff
@@ -568,8 +568,10 @@ angular.module('ngResource', ['ng']).
               undefined;
 
             forEach(action, function(value, key) {
-              if (key != 'params' && key != 'isArray' && key != 'interceptor') {
+              if (key != 'params' && key != 'isArray' && key != 'interceptor' && key != 'timeout') {
```
Comment (caitp) [inline src/ngResource/resource.js]: What if we just made this a switch statement for readability?

Comment (netman92) [inline src/ngResource/resource.js]: @caitp Done

**Comment by jtheoof** [comment]
:+1:

File: test/ngResource/resourceSpec.js [hunk 2]
```diff
@@ -1356,5 +1357,36 @@ describe('resource', function() {
       );
   });
 
+  it('If timeout promise is resolved, cancel the request', function() {
```
Comment (caitp) [inline test/ngResource/resourceSpec.js]: nit: can you change the test description to match the grammar of other ones in the file?

**Comment by caitp** [comment]
Other than the nit it looks great to me. Would be good if @gkalpak or @Narretz would take a look as well before merging

File: test/ngResource/resourceSpec.js (hunk 2 above)
Comment (netman92) [inline test/ngResource/resourceSpec.js]: @caitp done

File: test/ngResource/resourceSpec.js [hunk 3]
```diff
@@ -1356,5 +1357,36 @@ describe('resource', function() {
...
+      query: {
+        method: 'GET',
+        timeout: canceler.promise
+      }
+    });
+
+    CreditCard.query();
+
+    canceler.resolve();
+    expect(function() { $httpBackend.flush();}).toThrow(new Error("No pending request to flush !"));
```
Comment (gkalpak) [inline test/ngResource/resourceSpec.js]: Nit: `expect($httpBackend.flush).toThrow(...);` is more concise and readable (here and below).
(Else you need a space after `flush();`.)

Comment (caitp) [inline test/ngResource/resourceSpec.js]: If it (flush) uses `this` at all, it's not an option anyways

File: test/ngResource/resourceSpec.js [hunk 4]
```diff
@@ -1356,5 +1357,36 @@ describe('resource', function() {
...
+    canceler = $q.defer();
+    CreditCard = $resource('/CreditCard', {}, {
+      query: {
+        method: 'GET',
+        timeout: canceler.promise
+      }
+    });
+
+    CreditCard.query();
+    expect(function() { $httpBackend.flush();}).not.toThrow(new Error("No pending request to flush !"));
```
Comment (gkalpak) [inline test/ngResource/resourceSpec.js]: I would leave it at `.not.toThrow()` (to account for the unlikely situation where the error message changes and the behaviour breaks (simultaneously)).
It's also more concise and readable.

File: test/ngResource/resourceSpec.js (hunk 3 above)
Comment (netman92) [inline test/ngResource/resourceSpec.js]: @gkalpak fixed

File: test/ngResource/resourceSpec.js [hunk 5]
```diff
@@ -1356,5 +1357,36 @@ describe('resource', function() {
...
+    canceler = $q.defer();
+    CreditCard = $resource('/CreditCard', {}, {
+      query: {
+        method: 'GET',
+        timeout: canceler.promise
+      }
+    });
+
+    CreditCard.query();
+    expect($httpBackend.flush).not.toThrow(new Error("No pending request to flush !"));
```
Comment (gkalpak) [inline test/ngResource/resourceSpec.js]: I would leave it at `not.toThrow()` (accounting for the unlikely situation where the error message changes and the behaviour breaks (simultaneously)).
It's also more concise and readable.

Also, remove the 2 empty lines below.

Comment (netman92) [inline test/ngResource/resourceSpec.js]: @gkalpak Fixed

**Comment by gkalpak** [comment]
LGTM :+1: 
<sub>
TBH, I don't find it very useful, since having to recreate the resource class everytime a timeout promise is resolved doesn't seem very practical. But if people are using it, it's fine by me :smiley:
</sub>

**Comment by netman92** [comment]
@caitp Can you merge it?

**Comment by Narretz** [comment]
What @gkalpak said seems like the behavior is not well defined. Do we need to update the docs? Also, the commits must be squashed and the commit message doesn't follow our guidelines.

**Comment by caitp** [comment]
the behaviour seems pretty well defined to me @Narretz. It would be cool if you could override it on a per-request basis, but hey.

Some time ago, I had proposed that we allow letting you specify a factory function to produce the timeout promise, I can't remember if that's been done yet or not. That would solve the usability concerns.

@netman92 can you update the docs to indicate that the timeout parameter can be specified during initial resource class construction? Then I'll squash and merge

**Comment by netman92** [comment]
@caitp https://docs.angularjs.org/api/ngResource/service/$resource
and param timeout

```
timeout – {number|Promise} – timeout in milliseconds, or promise that should abort the request when resolved.
```
 I think, it is already in docs.

**Comment by caitp** [comment]
Not really, it's not an accurate depiction of what happens

**Comment by netman92** [comment]
@caitp Sorry, but I do not know, what to add to docs.

**Comment by jtheoof** [comment]
I agree with @netman92. The documentation seems correct, the behaviour was incorrect in a sense that the `Promise` was copied, therefore not cancellable. I think it's good to merge.

**Comment by jvorcak** [comment]
The documentation seems correct to me as well. Do you think it will be merged eventually @caitp? I'd really need this for one of my apps.

**Comment by gkalpak** [comment]
You know what they say: It's never too late, right ?)
Well, I just realized there is an issue with this change:

If the user fails to re-instantiate the resource after having cancelled a request (i.e. after having resolved the `timeout` promise), **all** subsequent requests will fail (and this behaviour isn't even documented).
You can see it in action [here][1] (with the current master).

IMO, we should come up with a better approach to request cancelling for `$resource` (let's discuss in #13058).
At the very least, a warning should be added in the docs and a BC notice in the changelog (but I would prefer fixing the behaviour instead). 

[1]: http://codepen.io/ExpertSystem/pen/meLoqE?editors=001

**Comment by petebacondarwin** [comment]
I agree - let's expedite #13058

**Comment by gkalpak** [comment]
#13058 has been superseeded by #13210.

**Pull request closed without merging**